src/tools/rbd_mirror/Replayer.cc

```diff
@@ -0,0 +1,125 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/debug.h"
+#include "common/errno.h"
+#include "include/stringify.h"
+#include "Replayer.h"
+
+#define dout_subsys ceph_subsys_rbd_mirror
+#undef dout_prefix
+#define dout_prefix *_dout << "rbd-mirror: "
+
+using std::map;
+using std::string;
+using std::unique_ptr;
+using std::vector;
+
+namespace rbd {
+namespace mirror {
+
+Replayer::Replayer(RadosRef primary_cluster, const peer_t &peer) :
+  m_lock(stringify("rbd::mirror::Replayer ") + stringify(peer)),
+  m_peer(peer),
+  m_primary(primary_cluster)
+{
+}
+
+Replayer::~Replayer()
+{
+}
+
+int Replayer::start()
+{
+  int r = m_remote->init2(m_peer.client_name.c_str(),
+			  m_peer.cluster_name.c_str(), 0);
+  if (r < 0) {
+    derr << "error initializing remote cluster for " << m_peer
+	 << " : " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  r = m_remote->conf_read_file(nullptr);
+  if (r < 0) {
+    derr << "could not read ceph conf for " << m_peer
+	 << " : " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  r = m_remote->connect();
+  if (r < 0) {
+    derr << "error connecting to remote cluster " << m_peer
+	 << " : " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  string cluster_uuid;
+  r = m_remote->cluster_fsid(&cluster_uuid);
+  if (r < 0) {
+    derr << "error reading cluster uuid from remote cluster " << m_peer
+	 << " : " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  if (cluster_uuid != m_peer.cluster_uuid) {
+    derr << "configured cluster uuid does not match actual cluster uuid. "
+	 << "expected: " << m_peer.cluster_uuid
+	 << " observed: " << cluster_uuid << dendl;
+    return -EINVAL;
+  }
+
+  return 0;
+}
+
+void Replayer::stop()
+{
+  m_remote->shutdown();
+}
+
+void Replayer::set_sources(const map<int64_t, set<string> > &images)
```

Comment:
I think this is backwards -- unless the idea is for the mirroring daemon to push updates to the remote.  You scan your local pools to know which pools are potentially mirrrored.   Then, for each remote pool, you would scan those pools for images that you need to sync/pull to the local pool.

This gets me thinking about how we should handle image removals -- and how to propagate those changes.  If we had an image being mirrored and we disable mirroring within one cluster, should the image be erased from the peers?  If the user deletes a mirrorred image from one cluster, should the image be erased from all peers?  Sounds like we almost need a journal for ```rbd_directory``` to track changes like that.  

And given that, should we just move the mirror enabled flag to the ```rbd_directory``` itself instead of storing it in each image?